applications/CompressiblePotentialFlowApplication/custom_strategies/transonic_residualbased_incrementalupdate_static_scheme.h

```diff
@@ -0,0 +1,343 @@
+
+    /**
+     * @brief Create method
+     * @param ThisParameters The configuration parameters
+     */
+    typename BaseSchemeType::Pointer Create(Parameters ThisParameters) const override
+    {
+        return Kratos::make_shared<ClassType>(ThisParameters);
+    }
+
+    /**
+     * @brief This is the place to initialize the Scheme.
+     * @details This is intended to be called just once when the strategy is initialized
+     * @param rModelPart The model part of the problem to solve
+     */
+    void Initialize(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mCriticalMach;
+        rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mUpwindFactorConstant;
+        rModelPart.GetProcessInfo()[MACH_LIMIT]             = std::sqrt(mMachNumberSquaredLimit);
+
+        BaseType::SetSchemeIsInitialized(true);
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief Function to be called when it is needed to initialize an iteration. It is designed to be called at the beginning of each non linear iteration
+     * @note Take care: the elemental function with the same name is NOT called here.
+     * @warning Must be defined in derived classes
+     * @details The function is called in the builder for memory efficiency
+     * @param rModelPart The model part of the problem to solve
+     * @param A LHS matrix
+     * @param Dx Incremental update of primary variables
+     * @param b RHS Vector
+     */
+    void InitializeNonLinIteration(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY
+
+        // Update the upwind factor constant and critical mach with the user-defined values
+        if (rModelPart.GetProcessInfo()[CONVERGENCE_RATIO] <= mUpdateRelativeResidualNorm &&
+            rModelPart.GetProcessInfo()[NL_ITERATION_NUMBER] > 1                          &&
+            mUpdatedValues == false){
+
+            rModelPart.GetProcessInfo()[CRITICAL_MACH]          = mNewCriticalMach;
+            rModelPart.GetProcessInfo()[UPWIND_FACTOR_CONSTANT] = mNewUpwindFactorConstant;
+
+            mUpdatedValues = true;
+        }
+
+        // Initialize non-linear iteration for all of the elements, conditions and constraints
+        EntitiesUtilities::InitializeNonLinearIterationAllEntities(rModelPart);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This method provides the defaults parameters to avoid conflicts between the different constructors
+     * @return The default parameters
+     */
+    Parameters GetDefaultParameters() const override
+    {
+        Parameters default_parameters = Parameters(R"(
+        {
+            "critical_mach"                 : 0.92,
+            "upwind_factor_constant"        : 2.0,
+            "new_critical_mach"             : 0.92,
+            "new_upwind_factor_constant"    : 2.0,
```

Comment:
I am not a big fan of the prefix "new" as it is not so clear what it's the purpose of the settings.

These are just two suggestions:

I personally prefer to define an "initial" set of parameters:
```suggestion
            "inital_critical_mach"                 : 0.92,
            "initial_upwind_factor_constant"        : 2.0,
            "critical_mach"             : 0.92,
            "upwind_factor_constant"    : 2.0,
```

Or alternatively, define the final values as "target"

```suggestion
            "critical_mach"                 : 0.92,
            "upwind_factor_constant"        : 2.0,
            "target_critical_mach"             : 0.92,
            "target_upwind_factor_constant"    : 2.0,
```

Reply:
Also, note that you are not doing any check on the initial/final settings. What happens if the user sets a "new" value lower than the initial one?

Reply:
Hi Marc, thanks! yes, I tried to change it in a linear way and the same with the critical mach but I didn't achieve good results. I think I can try it again, just to try.

Related to the names, my intention was just to make it clear that the initial parameters are the same as those in the ApplyFarFieldProcess, here the scheme is overriting it. I think the two prefixes (initial_ and target_) could be added to make it clearer.

In the case of the parameter values, it is expected a target_critical_mach greater than the initial_critical_mach (and less than 0.99). In the case of the constant, its target_value is expected to be less than the initial one. If the initial and target values are the same, it would be like not making the change, so I leave them that way by default.

Since They are user-defined, I think It is better not to restrict it too much to be able to test values ​​for each particular case.